### PR TITLE
[TACHYON-984] Reenable raw table master client related API and tests

### PR DESCRIPTION
**Pull request by uronce-cc**
https://tachyon.atlassian.net/browse/TACHYON-984

Reopen https://github.com/amplab/tachyon/pull/1594 with `retryRpc` limited to `RawTableMasterClient`.

### CODE REVIEW

**Comment by danielfree** [comment]
Is it possible to move these methods into new TachyonFileSystem api instead of TachyonFS?

**Comment by uronce-cc** [comment]
@danielfree I just planned to reenable the original API, even if we want new API for raw table client, we should have a separate client like `TachyonRawTable` other than adding to `TachyonFileSystem`.

**Comment by uronce-cc** [comment]
@calvinjia PTAL

File: common/src/main/java/tachyon/client/RawTableMasterClient.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+   * @throws IOException when retries exceeds {@link #RPC_MAX_NUM_RETRY} or {@link #close()} has
+   *                     been called before calling this method or during the retry
+   */
+  private <V> V retryRPC(RpcCallable<V> rpc) throws IOException {
+    int retry = 0;
+    while (!mClosed && (retry ++) <= RPC_MAX_NUM_RETRY) {
+      connect();
+      try {
+        return rpc.call();
+      } catch (TException e) {
```
Comment (calvinjia) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: I think one limitation is we cannot catch specific exceptions anymore

File: common/src/main/java/tachyon/client/RawTableMasterClient.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+  /**
+   * Creates a new raw table master client.
+   *
+   * @param masterAddress the master address
+   * @param executorService the executor service
+   * @param tachyonConf the Tachyon configuration
+   */
+  public RawTableMasterClient(InetSocketAddress masterAddress, ExecutorService executorService,
+                              TachyonConf tachyonConf) {
```
Comment (calvinjia) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: Indentation

**Comment by calvinjia** [comment]
@uronce-cc Thanks for the patch, could you modify the bin/tachyon runTests to also do the raw table logic?

File: common/src/main/java/tachyon/client/RawTableMasterClient.java (hunk 1 above)
Comment (uronce-cc) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: Yes, that's why the solution is not best, so I limited its use in this class.

**Comment by uronce-cc** [comment]
@calvinjia PTAL, Thanks.

File: common/src/main/java/tachyon/client/RawTableMasterClient.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    return Constants.RAW_TABLE_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  protected void afterConnect() {
+    mClient = new RawTableMasterService.Client(mProtocol);
+  }
+
+  /**
+   * Create a raw table. A table is a directory with sub-directories representing columns.
```
Comment (jsimsa) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: ``Creates a raw table.``

File: common/src/main/java/tachyon/client/RawTableMasterClient.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    return retryRPC(new RpcCallable<Long>() {
+      @Override
+      public Long call() throws TException {
+        return mClient.createRawTable(path.getPath(), columns, metadata);
+      }
+    });
+  }
+
+  /**
+   * Get the {@link RawTableInfo} associated with the given id.
```
Comment (jsimsa) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: ``Gets ...``

File: common/src/main/java/tachyon/client/RawTableMasterClient.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    return retryRPC(new RpcCallable<RawTableInfo>() {
+      @Override
+      public RawTableInfo call() throws TException {
+        return mClient.getClientRawTableInfoById(id);
+      }
+    });
+  }
+
+  /**
+   * Get the {@link RawTableInfo} associated with the given path.
```
Comment (jsimsa) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: ``Gets ...``

File: common/src/main/java/tachyon/client/RawTableMasterClient.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    return retryRPC(new RpcCallable<RawTableInfo>() {
+      @Override
+      public RawTableInfo call() throws TException {
+        return mClient.getClientRawTableInfoByPath(path.getPath());
+      }
+    });
+  }
+
+  /**
+   * Update the metadata of a table.
```
Comment (jsimsa) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: ``Updates ...``

File: common/src/main/java/tachyon/client/RawTableMasterClient.java (hunk 1 above)
Comment (jsimsa) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: As per the exception handling proposal, the Thrift RPC calls should only throw TException and TachyonTException. We throw the TachyonTException to the caller who can take care of the unwrapping.

File: integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java [hunk 7]
```diff
@@ -39,16 +51,15 @@ public final void after() throws Exception {
   public final void before() throws Exception {
     mLocalTachyonCluster = new LocalTachyonCluster(10000, 1000, Constants.GB);
     mLocalTachyonCluster.start();
-    mTfs = mLocalTachyonCluster.getClient();
+    mTfs = mLocalTachyonCluster.getOldClient();
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/table/RawTableIntegrationTest.java]: Why are we switching from the new client to the old client? That seems backwards.

File: integration-tests/src/test/java/tachyon/client/table/RawColumnIntegrationTest.java [hunk 8]
```diff
@@ -27,21 +35,26 @@
...
 public class RawColumnIntegrationTest {
   private LocalTachyonCluster mLocalTachyonCluster = null;
-  private TachyonFileSystem mTfs = null;
+  private TachyonFS mTfs = null;
+
+  @Before
+  public final void before() throws Exception {
+    mLocalTachyonCluster = new LocalTachyonCluster(10000, 1000, Constants.GB);
+    mLocalTachyonCluster.start();
+    mTfs = mLocalTachyonCluster.getOldClient();
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/table/RawColumnIntegrationTest.java]: Why are we switching from the new client to the old client. That seems backwards?

**Comment by jsimsa** [comment]
Jenkins, test this please.

**Comment by jsimsa** [comment]
Left some comments for you.

File: common/src/main/java/tachyon/client/RawTableMasterClient.java (hunk 1 above)
Comment (uronce-cc) [inline common/src/main/java/tachyon/client/RawTableMasterClient.java]: Yeah, that's one solution, but any way, somebody has to do the unwrapping.

File: integration-tests/src/test/java/tachyon/client/table/RawColumnIntegrationTest.java (hunk 8 above)
Comment (uronce-cc) [inline integration-tests/src/test/java/tachyon/client/table/RawColumnIntegrationTest.java]: The raw table client is only supported in old client API.

**Comment by uronce-cc** [comment]
@calvinjia @jsimsa Comments resolved, PTAL.

**Comment by jsimsa** [comment]
LGTM +1

**Comment by calvinjia** [comment]
LGTM

**Pull request merged**